### PR TITLE
xml-SAX and DOM parsers

### PR DESCRIPTION
**Pull request by slavidichkov**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,10 @@
 *.class
+out/
+src/com/clouway/practice/
+src/com/clouway/practice/*
 
+.idea/workspace.xml
+.idea/vcs.xml
```
Comment (tsony-tsonev) [inline .gitignore]: You have to ignore the whole .idea directory. I see you have committed a lot of files from there which should not be part of the PR. You have to remove them.

File: src/com/clouway/task/Address.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    public String getSection() {
+        return section;
+    }
+
+    public String getCity() {
+        return city;
+    }
+
+    @Override
+    public String toString() {
```
Comment (tsony-tsonev) [inline src/com/clouway/task/Address.java]: don't override toString if you are not going to use it.

File: src/com/clouway/task/Employee.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    public Employee(String firstName, String lastName, String position, int age) {
+        this.firstName = firstName;
+        this.lastName = lastName;
+        this.position = position;
+        this.age = age;
+    }
+
+    @Override
+    public String toString() {
```
Comment (tsony-tsonev) [inline src/com/clouway/task/Employee.java]: same here

File: src/com/clouway/task/Employee.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    public int addressesListSize() {
+        return addresses.size();
+    }
+
+    public void printContentOfEmployersList() {
+       for (int i=0;i<employers.size();i++){
+           System.out.println(employers.get(i).getName());
+//           System.out.println(employers.get(i).getStartDate());
+//           System.out.println(employers.get(i).getEndDate());
```
Comment (tsony-tsonev) [inline src/com/clouway/task/Employee.java]: commented code, also this printing methods printContentOfEmployersList, printContentOfAddressesList should not be part of your object

File: tests/com/clouway/task/DomParserTest.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    @Before
+    public void setUp(){
+        inputStream=new ByteArrayInputStream(xmlFile.getBytes(StandardCharsets.UTF_8));
+    }
+
+    @Test
+    public void pretendSameFirstNameOfInstance() {
+        DomParser<Employee> employeeDomParser = new DomParser<>();
+        List<Employee> employees = employeeDomParser.parse(Employee.class, inputStream);
+        assertThat(employees.get(0).getFirstName(),is(equalTo("Ivan")));
```
Comment (tsony-tsonev) [inline tests/com/clouway/task/DomParserTest.java]: you should assert that all fields are as expected not just one

File: tests/com/clouway/task/SAXParserTest.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        employeeList=saxParser.parse(Employee.class,inputStream);
+        assertThat(employeeList.size(),is(equalTo(2)));
+    }
+
+    @Test
+    public void pretendSameFirstNameOfInstance() {
+        SaxParser<Employee> saxParser=new SaxParser<>();
+        List<Employee> employeeList;
+        employeeList=saxParser.parse(Employee.class,inputStream);
+        assertThat(employeeList.get(0).getFirstName(),is(equalTo("Ivan")));
```
Comment (tsony-tsonev) [inline tests/com/clouway/task/SAXParserTest.java]: same here

File: xml.iml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module type="JAVA_MODULE" version="4">
+  <component name="NewModuleRootManager" inherit-compiler-output="true">
+    <exclude-output />
```
Comment (tsony-tsonev) [inline xml.iml]: this file also should not be part of this PR

File: src/com/clouway/task/DomParser.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+import java.util.Collection;
+import java.util.Date;
+import java.util.List;
+
+/**
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public class DomParser<T> implements XmlParser<T> {
+    private InputStream inputStream;
+    private DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
```
Comment (tsony-tsonev) [inline src/com/clouway/task/DomParser.java]: this could be initialized in empty constructor

File: src/com/clouway/task/DomParser.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    private DocumentBuilder dBuilder = null;
+    private Document doc = null;
+
+    public void setValidating(boolean validating){
+        if (validating){
+            dbFactory.setValidating(true);
+        }else {
+            dbFactory.setValidating(false);
+        }
```
Comment (tsony-tsonev) [inline src/com/clouway/task/DomParser.java]: you could use just 

```java
dbFactory.setValidating(validating);
```
instead if else, also I'm not sure this method is used at all

File: src/com/clouway/task/DomParser.java [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        } catch (ParserConfigurationException e) {
+            e.printStackTrace();
+        } catch (SAXException e) {
+            e.printStackTrace();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        NodeList nList = doc.getElementsByTagName(aClass.getSimpleName());
+
+        return (List<T>) parse(nList, aClass);
```
Comment (tsony-tsonev) [inline src/com/clouway/task/DomParser.java]: you can make the private method return List<T> to prevent this casting

File: src/com/clouway/task/DomParser.java [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    private List<Object> parse(NodeList childNodes, Class aClass) {
+        List<Object> instanceList = new ArrayList<>();
```
Comment (tsony-tsonev) [inline src/com/clouway/task/DomParser.java]: don't add the collection type in the variable name just "instances" is enaugh

File: src/com/clouway/task/DomParser.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+        } catch (IllegalAccessException e) {
+            e.printStackTrace();
+        }
+        for (int i = 0; i < fields.length; i++) {
+            setField(fields[i], instance, element, parentsNumber);
+        }
+        return instance;
+    }
+
+    private List<Object> createChildesList(NodeList nodeList, Class aClass, int parentsNumber) {
```
Comment (tsony-tsonev) [inline src/com/clouway/task/DomParser.java]: just createChildren

File: src/com/clouway/task/DomParser.java [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+                Element eElement = (Element) node;
+                instance=createObject(aClass,eElement,parentsNumber);
+                instanceList.add(instance);
+            }
+            node = node.getNextSibling();
+        }
+        return instanceList;
+    }
+
+    private String getValue(Field field, Element element) {
```
Comment (tsony-tsonev) [inline src/com/clouway/task/DomParser.java]: getElementValue

File: src/com/clouway/task/Address.java (hunk 2 above)
Comment (slavidichkov) [inline src/com/clouway/task/Address.java]: removed toString

File: src/com/clouway/task/Employee.java (hunk 3 above)
Comment (slavidichkov) [inline src/com/clouway/task/Employee.java]: done

File: src/com/clouway/task/Employee.java (hunk 4 above)
Comment (slavidichkov) [inline src/com/clouway/task/Employee.java]: I removed it

File: src/com/clouway/task/DomParser.java (hunk 9 above)
Comment (slavidichkov) [inline src/com/clouway/task/DomParser.java]: fixed

File: src/com/clouway/task/DomParser.java (hunk 8 above)
Comment (slavidichkov) [inline src/com/clouway/task/DomParser.java]: fixed

File: src/com/clouway/task/DomParser.java (hunk 12 above)
Comment (slavidichkov) [inline src/com/clouway/task/DomParser.java]: fixed

File: src/com/clouway/task/DomParser.java (hunk 11 above)
Comment (slavidichkov) [inline src/com/clouway/task/DomParser.java]: fixed

File: src/com/clouway/task/DomParser.java (hunk 13 above)
Comment (slavidichkov) [inline src/com/clouway/task/DomParser.java]: done

File: tests/com/clouway/task/DomParserTest.java (hunk 5 above)
Comment (slavidichkov) [inline tests/com/clouway/task/DomParserTest.java]:
```assertThat(employees.get(0).getFirstName(),is(equalTo("Ivan")));
        assertThat(employees.get(0).getLastName(),is(equalTo("Ivanov")));
        assertThat(employees.get(0).getPosition(),is(equalTo("mechanic")));
        assertThat(employees.get(0).getAge(),is(equalTo(23)));```

Comment (tsony-tsonev) [inline tests/com/clouway/task/DomParserTest.java]: if theese are all the fileds yep, you could also override equals and hash code and assert just the objects not field by field

File: xml.iml (hunk 7 above)
Comment (slavidichkov) [inline xml.iml]: fixed

File: src/com/clouway/task/DomParser.java (hunk 10 above)
Comment (slavidichkov) [inline src/com/clouway/task/DomParser.java]: fixed

File: .gitignore (hunk 1 above)
Comment (slavidichkov) [inline .gitignore]: done

File: tests/com/clouway/task/DomParserTest.java (hunk 5 above)
Comment (slavidichkov) [inline tests/com/clouway/task/DomParserTest.java]: fixed

File: tests/com/clouway/task/SAXParserTest.java (hunk 6 above)
Comment (slavidichkov) [inline tests/com/clouway/task/SAXParserTest.java]: fixed

**Comment by slavidichkov** [comment]
all comments  fixed

**Comment by tsony-tsonev** [comment]
LGTM

File: src/com/clouway/task/DomParser.java [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    private InputStream inputStream;
+    private DocumentBuilderFactory dbFactory;
+    private DocumentBuilder dBuilder = null;
+    private Document doc = null;
+
+    public DomParser(){
+        dbFactory = DocumentBuilderFactory.newInstance();
+    }
+
+    public void setValidating(boolean validating){
```
Comment (mgenov) [inline src/com/clouway/task/DomParser.java]: Why this configuration is exposed with setter ?

File: src/com/clouway/task/DomParser.java [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+            node = node.getNextSibling();
+        }
+        return instances;
+    }
+
+    private String getElementValue(Field field, Element element) {
+        return element.getElementsByTagName(field.getName()).item(0).getTextContent();
+    }
+
+    public Date parseDate(String date) {
```
Comment (mgenov) [inline src/com/clouway/task/DomParser.java]: Why this method is visible from outside ?

File: src/com/clouway/task/SAXHandler.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Date;
+import java.util.List;
+
+/**
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public class SAXHandler<T> extends DefaultHandler {
```
Comment (mgenov) [inline src/com/clouway/task/SAXHandler.java]: Why this object is with public visibility ?

File: src/com/clouway/task/XmlParser.java [hunk 17]
```diff
@@ -0,0 +1,11 @@
+package com.clouway.task;
+
+import java.io.InputStream;
+import java.util.List;
+
+/**
+ * @author Slavi Dichkov (slavidichkof@gmail.com)
+ */
+public interface XmlParser<T> {
```
Comment (mgenov) [inline src/com/clouway/task/XmlParser.java]: Why definition of this interface says that I have to create parser instance for each object that I have to parse?

File: tests/com/clouway/task/DomParserTest.java [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+        employers.add(new Employer("Stefan", getDate("2010-10-21"),getDate("2012-4-23")));
+        employers.add(new Employer("Ivan",  getDate("2010-10-21"),getDate("2012-4-23")));
+        assertThat(employees.get(0).getEmployers(),is(equalTo(employers)));
+
+        List<Address> addresses = new ArrayList();
+        addresses.add(new Address("Gabrovski",43,"A","Veliko Tyrnovo"));
+        assertThat(employees.get(0).getAddresses(),is(equalTo(addresses)));
+    }
+
+    private Date getDate(String date){
```
Comment (mgenov) [inline tests/com/clouway/task/DomParserTest.java]: Why this code is duplicate on 3 places ? I think that I saw it in SAX parser, DOM parser and in tests.

Comment (slavidichkov) [inline tests/com/clouway/task/DomParserTest.java]: There is no reason I could extract it in a separate method assertThatTwoListsAreTheSame(list1,list2)

Comment (mgenov) [inline tests/com/clouway/task/DomParserTest.java]: I don't get how this will remove duplication from all of them ?

Comment (slavidichkov) [inline tests/com/clouway/task/DomParserTest.java]: I had to extract it in a separate class where getDate(String date) method had to be

File: src/com/clouway/task/DomParser.java (hunk 15 above)
Comment (slavidichkov) [inline src/com/clouway/task/DomParser.java]: I forgot to make it private

File: src/com/clouway/task/XmlParser.java (hunk 17 above)
Comment (slavidichkov) [inline src/com/clouway/task/XmlParser.java]: I can make the method generic not the whole class. By this way i wold be able to parse different  XMLs with one instance

Comment (mgenov) [inline src/com/clouway/task/XmlParser.java]: Sounds. The question is why it wasn't designed like that at first place.

Comment (mgenov) [inline src/com/clouway/task/XmlParser.java]: Sounds good*

File: src/com/clouway/task/SAXHandler.java (hunk 16 above)
Comment (slavidichkov) [inline src/com/clouway/task/SAXHandler.java]: I can move it like a private inner class in SAXParser to be visible only for SAXParser

File: src/com/clouway/task/XmlParser.java (hunk 17 above)
Comment (slavidichkov) [inline src/com/clouway/task/XmlParser.java]: fixed

File: src/com/clouway/task/SAXHandler.java (hunk 16 above)
Comment (slavidichkov) [inline src/com/clouway/task/SAXHandler.java]: i made it inner private class
